src/pybind/mgr/cephadm/schedule.py

```diff
@@ -0,0 +1,183 @@
+
+        chosen = hosts
+
+
+        chosen, existing = self.prefer_hosts_with_daemons(chosen)
+
+        need = count - len(existing)
+        others = difference_hostspecs(chosen, existing)
+
+        if need < 0:
+            return self.scheduler.place(existing, count)
+        else:
+            if self.filter_new_host:
+                old = others
+                others = [h for h in others if self.filter_new_host(h.hostname)]
+                logger.debug('filtered %s down to %s' % (old, hosts))
+
+            chosen = self.scheduler.place(others, need)
+            logger.debug('Combine hosts with existing daemons %s + new hosts %s' % (
+                existing, chosen))
+            return list(merge_hostspecs(existing, chosen))
+
+    def get_candidates(self) -> List[HostPlacementSpec]:
+        if self.spec.placement.hosts:
+            return self.spec.placement.hosts
+        elif self.spec.placement.label:
+            return [
+                HostPlacementSpec(x, '', '')
+                for x in self.get_hosts_func(self.spec.placement.label)
+            ]
+        elif self.spec.placement.host_pattern:
+            return [
+                HostPlacementSpec(x, '', '')
+                for x in self.spec.placement.pattern_matches_hosts(self.get_hosts_func(None))
+            ]
+        if self.spec.placement.count is None:
+            raise OrchestratorValidationError("placement spec is empty: no hosts, no label, no pattern, no count")
+        # backward compatibility: consider an empty placements to be the same pattern = *
+        return [
+            HostPlacementSpec(x, '', '')
+            for x in self.get_hosts_func(None)
+        ]
+
+    def prefer_hosts_with_daemons(self, chosen: List[HostPlacementSpec]) -> Tuple[List[HostPlacementSpec], List[HostPlacementSpec]]:
+        # prefer hosts that already have services
+        daemons = self.get_daemons_func(self.service_name)
+        hosts_with_daemons = {d.hostname for d in daemons}
+        # calc existing daemons (that aren't already in chosen)
+        existing = [hs for hs in chosen if hs.hostname in hosts_with_daemons]
+        count = self.spec.placement.count
+        assert count is not None
+        if len(list(merge_hostspecs(chosen, existing))) >= count:
+            chosen = list(merge_hostspecs(chosen, self.scheduler.place(
+                existing,
+                count - len(chosen))))
+            logger.debug('Hosts with existing daemons: {}'.format(chosen))
+        return chosen, existing
+
+
+def merge_hostspecs(l: List[HostPlacementSpec], r: List[HostPlacementSpec]) -> Iterable[HostPlacementSpec]:
+    l_names = {h.hostname for h in l}
+    yield from l
+    yield from (h for h in r if h.hostname not in l_names)
+
+
+def difference_hostspecs(l: List[HostPlacementSpec], r: List[HostPlacementSpec]) -> List[HostPlacementSpec]:
+    r_names = {h.hostname for h in r}
+    return [h for h in l if h.hostname not in r_names]
```

Comment:
This whole section needs docstrings, descriptive variable names and a bit of de-tangling.